### PR TITLE
[SPARK-31000][PYTHON][SQL] Add ability to set table description via Catalog.createTable()

### PR DESCRIPTION
**Pull request by nchammas**
### What changes were proposed in this pull request?

This PR enhances `Catalog.createTable()` to allow users to set the table's description. This corresponds to the following SQL syntax:

```sql
CREATE TABLE ...
COMMENT 'this is a fancy table';
```

### Why are the changes needed?

This brings the Scala/Python catalog APIs a bit closer to what's already possible via SQL.

### Does this PR introduce any user-facing change?

Yes, it adds a new parameter to `Catalog.createTable()`.

### How was this patch tested?

Existing unit tests:

```sh
./python/run-tests \
  --python-executables python3.7 \
  --testnames 'pyspark.sql.tests.test_catalog,pyspark.sql.tests.test_context'
```

```
$ ./build/sbt
testOnly org.apache.spark.sql.internal.CatalogSuite org.apache.spark.sql.CachedTableSuite org.apache.spark.sql.hive.MetastoreDataSourcesSuite org.apache.spark.sql.hive.execution.HiveDDLSuite
```

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/CachedTableSuite.scala
Comment (nchammas) [inline sql/core/src/test/scala/org/apache/spark/sql/CachedTableSuite.scala]: I had to make these changes so that the correct method would be invoked. I'm not fluent in Scala, but at least in Python I'd consider this style to also be better form since explicitly naming the parameters eliminates ambiguity.

Would this be considered an API breaking change, though? Since folks who are using the method with unnamed parameters will have trouble once `description` gets added in.

If so, would the solution be to move `description` to the end of the parameter list?

Comment (nchammas) [reply sql/core/src/test/scala/org/apache/spark/sql/CachedTableSuite.scala]: > This patch fails MiMa tests.

I guess that's a yes, this breaks binary compatibility. Let me try moving the parameter around.

**Comment by nchammas** [comment]
OK, I think I've addressed the binary compatibility issues, but as a Scala n00b I'm a little surprised at the size of the diff required to add a single parameter.

Is there a better way to do what I'm trying to do?

**Comment by nchammas** [comment]
Jenkins, my man, retest this please.

**Comment by nchammas** [comment]
Retest this please.

**Comment by nchammas** [comment]
@HyukjinKwon - I noticed you added yourself as a watcher on SPARK-31000, so perhaps you'd be interested in this PR. It's my [first time](https://github.com/apache/spark/pull/27908#issuecomment-599020089) changing something on the Scala side of things.

Am I approaching this the right way?

File: sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala:370]: I guess the version has to be changed to `@since 3.1.0`

File: python/pyspark/sql/catalog.py
Comment (HyukjinKwon) [inline python/pyspark/sql/catalog.py:164]: Preferably I would add `.. versionchanged:: 3.1`.

**Comment by HyukjinKwon** [comment]
What's your thought @hvanhovell and @cloud-fan? Git blame says you guys touched the codes most recently.

**Comment by rberenguel** [comment]
The changes look reasonable to me, it doesn't seem to add anything public (except on the Python side, where that _should_ be fine) and the tests are there (also, I'd use this 😄 )

**Comment by nchammas** [comment]
Thanks for taking a look, @HyukjinKwon and @rberenguel.

> The changes look reasonable to me, it doesn't seem to add anything public (except on the Python side, where that should be fine)

There are new public Scala APIs to allow use of the new `description` parameter, no?

**Comment by rberenguel** [comment]
> Thanks for taking a look, @HyukjinKwon and @rberenguel.
> 
> > The changes look reasonable to me, it doesn't seem to add anything public (except on the Python side, where that should be fine)
> 
> There are new public Scala APIs to allow use of the new `description` parameter, no?

Sorry, I checked the wrong view (mobile Github still leaves a bit to be desired). I think this is not a problem, because it's not modifying any existing public API, just adding to it. Updating the version tags for the API was needed indeed, 👍

**Comment by nchammas** [comment]
Jenkins, retest this please.

**Comment by nchammas** [comment]
Is this good to go?

**Comment by nchammas** [comment]
Checking in again here.

**Comment by nchammas** [comment]
Retest this please.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by nchammas** [comment]
Ayyo Jenkins, retest this please. And make it p a s s. 🕺

**Comment by nchammas** [comment]
@HyukjinKwon - This little feature is ready to go. Do we want to get input from another committer before merging it in?

**Review by hvanhovell — APPROVED** [review]
LGTM

**Comment by nchammas** [comment]
Thanks for taking a look, @hvanhovell.

File: python/pyspark/sql/catalog.py
Comment (HyukjinKwon) [inline python/pyspark/sql/catalog.py:163]: @nchammas, just for doubly sure, you tested this in PySpark manually, right?

**Review by HyukjinKwon — APPROVED** [review]

**Comment by nchammas** [comment]
Test failures seem unrelated to this PR.

Just to be sure: Jenkins! Retest this please.

**Comment by nchammas** [comment]
Retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala:375]: @nchammas, BTW why did you choose the name `description` over `comment`? Seems `comment` is more consistent.

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala:375]: Also, it's a slight behaviour change I guess. If there's no description set, I think we should set `None` instead of `Some("")`.

Comment (nchammas) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: Good point. Fixed. I made `description` an `Option[String]`. (I _think_ that's the right thing to do here...)

**Comment by nchammas** [comment]
```
py4j.protocol.Py4JError: An error occurred while calling o42.createTable. Trace:
py4j.Py4JException: Method createTable([class java.lang.String, class java.lang.String, class org.apache.spark.sql.types.StructType, class java.util.HashMap, class java.lang.String]) does not exist
```

I think I need to reorganize things a bit to get Python to work. py4j, I think, doesn't work with `Option` types. I believe this is one of the reasons I went with `String = ""` initially. Let me take another look. Perhaps I'm missing something simple.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala:328]: Oh yeah, we shouldn't use `Option` because this is a Scala specific collection. We will just have to override so Java side can call these as well.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: Hm.. okay, should be fine. At least it's consistent with one side.

Comment (nchammas) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: Just to be clear, are you saying to leave the name as-is in this PR, or to change it to `comment`?

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala
Comment (nchammas) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala:328]: I've switched to a plain `String` in the Java version of this API. (Not sure if I'm using the terminology correctly.)

File: sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala:358]: @nchammas, could we place `option` at the last?

File: sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala:381]: I think we don't need to make the `description` `Option`al. We can just call:

```scala
  override def createTable(
      tableName: String,
      source: String,
      schema: StructType,
      options: Map[String, String]): DataFrame = {
    createTable(tableName, source, schema, options, None)
  }
```

if `description` is not specified in Python side.

File: sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala:474]: Likewise, let's just expose `description` as a plain string argument.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala:375]: I think you can take `description` as `String` in the argument and here pass it via `Option(description)`.

File: sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala
Comment (nchammas) [reply sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala:358]: Can do, but just so I understand: Consistent with what?

File: sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala
Comment (nchammas) [reply sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala:381]: Python won't call this particular signature.

I assume you meant to make this comment on the signature that includes `java.util.Map[String, String]`, right? `description` is already a plain `String` there. The `Option[String]` is just for the Scala API.

Does that address your concern, or am I misunderstanding something?

Comment (nchammas) [reply sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala:381]: Just to be clear about what I think I'm doing here (because I'm unsure of myself and could use your confirmation):

At a high level, we have two versions of the `createTable()` APIs: one for Scala, and one for Java and Python. Python calls the Java versions of the API via Py4J.

On the Scala side, it's OK to use `Option[String]`. On the Java/Python side, we need to use a plain `String` and then convert that to an `Option[String]` when calling from the Java side into the Scala. That's what I'm currently doing.

Am I understanding correctly?

File: sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala:358]: I mean consistency with other APIs above which places `options` at the last.

File: sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala:381]: Hmmmm .. why should this be `Option` alone in Scala? I thought it's better to keep the signature close to each other.

**Comment by HyukjinKwon** [comment]
Ah, I think it's okay with `description`.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

Thanks @nchammas.

**Pull request closed without merging**
